File: src/class068/C1_DistinctSubsequences.cpp
#include <vector>
#include <string>

class Code01_DistinctSubsequences {
public:
    // 动态规划方法
    static int numDistinct1(const std::string& str, const std::string& target) {
        int n = str.size();
        int m = target.size();
        std::vector<std::vector<int>> dp(n + 1, std::vector<int>(m + 1, 0));

        // 初始化，s的任何前缀形成的序列都包含空序列
        for (int i = 0; i <= n; i++) {
            dp[i][0] = 1;
        }

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                dp[i][j] = dp[i - 1][j]; // 不使用s中的当前字符
                if (str[i - 1] == target[j - 1]) {
                    dp[i][j] += dp[i - 1][j - 1]; // 使用s中的当前字符
                }
            }
        }
        return dp[n][m];
    }

    // 动态规划，空间优化版
    static int numDistinct2(const std::string& str, const std::string& target) {
        int n = str.size();
        int m = target.size();
        std::vector<int> dp(m + 1, 0);
        dp[0] = 1; // 任何字符串的子序列都包含空序列

        for (int i = 1; i <= n; i++) {
            for (int j = m; j >= 1; j--) { // 从后向前更新，避免干扰未处理的数据
                if (str[i - 1] == target[j - 1]) {
                    dp[j] += dp[j - 1]; // 更新dp[j]，使用当前s中的字符
                }
            }
        }
        return dp[m];
    }
};

File: src/class068/C1_DistinctSubsequences.py
class Code01_DistinctSubsequences:
    @staticmethod
    # 动态规划方法
    def numDistinct1(s, target):
        n = len(s)
        m = len(target)
        dp = [[0] * (m + 1) for _ in range(n + 1)]

        # 初始化，s的任何前缀形成的序列都包含空序列
        for i in range(n + 1):
            dp[i][0] = 1

        for i in range(1, n + 1):
            for j in range(1, m + 1):
                dp[i][j] = dp[i - 1][j]  # 不使用s中的当前字符
                if s[i - 1] == target[j - 1]:
                    dp[i][j] += dp[i - 1][j - 1]  # 使用s中的当前字符

        return dp[n][m]

    @staticmethod
    # 动态规划，空间优化版
    def numDistinct2(s, target):
        n = len(s)
        m = len(target)
        dp = [0] * (m + 1)
        dp[0] = 1  # 任何字符串的子序列都包含空序列

        for i in range(1, n + 1):
            for j in range(m, 0, -1):  # 从后向前更新，避免干扰未处理的数据
                if s[i - 1] == target[j - 1]:
                    dp[j] += dp[j - 1]  # 更新dp[j]，使用当前s中的字符

        return dp[m]
